### ISSUE TITLE
SearchBar Typescript types are broken

### ISSUE THREAD

**Issue by djMax** [react-native-elements/react-native-elements#3089]
**Explain what you did (Required)**

I'm using SearchBar in typescript. I have a state variable to store the search text:

```
  const [search, setSearch] = React.useState<string>();
  return (
      <SearchBar
          onChangeText={setSearch}
        />
  )
```

**Expected behavior (Required)**

I expect this to be a valid usage according to Typescript.

**Describe the bug (Required)**

TS says onChangeText doesn't match, but the required property type is essentially disjoint (I think).

```
index.tsx(74,11): error TS2322: Type '(text: string) => void' is not assignable to type '((text: string) => void) & ((text: string) => void) & (() => any) & (() => any) & (() => any) & ((text: string) => void) & (() => any) & (() => any) & (() => any)'.
  Type '(text: string) => void' is not assignable to type '() => any'.
```

**Your Environment (Required):**

| software              | version |
| --------------------- | ------- |
| react-native-elements |     3.4.2    |
| react-native          |     0.64.2    |

**Comment by rush86999** [comment]
i am having the same problem

**Comment by marcuiulian13** [comment]
To add to this, the `SearchBar` types seem to have been broken in the last release.

One can do `onChangeText={setSearch as any}` to get rid of the error initially reported here, but then Typescript complains that you're missing even more **required** properties, like `onBlur`, `onFocus`, `onClear`, etc. which should not always be implemented by the developer:
![image](https://user-images.githubusercontent.com/3667966/122679213-e7b24980-d1e1-11eb-8521-4bbce7b53179.png)

I checked the diff from last release and it seems like this change has broken the types: https://github.com/react-native-elements/react-native-elements/compare/v3.4.1...v3.4.2#diff-faa340aefe5bfd898992f880448fa1665c605da68118c860aaa29f55ebbc5bd7L48
Specifically, `SearchBar` has been changed to accept `SearchBarProps`, which is an union of `SearchBarDefaultProps` & `SearchBarAndroidProps` & `SearchBarIOSProps`, which are wrongly requiring all the different handlers instead of marking them as optional.

@djMax could you please update the title of this issue to reflect the fact that all types for `SearchBar` are broken?

**Comment by djMax** [comment]
Changed. As my teenager would say, lol xd bad.

**Comment by marcuiulian13** [comment]
For the meantime, my current solution for this regression is to use the `SearchBar` like this (overwriting the prop types to what was in v3.4.1) :
```
import { SearchBar } from 'react-native-elements';

const SafeSearchBar = (SearchBar as unknown) as React.FC<SearchBarBaseProps>;

....
<SafeSearchBar
  platform="default"
   ...
/>
```

**Comment by arpitBhalla** [comment]
```
export type SearchBarProps =
  | SearchBarBaseProps
  | SearchBarDefaultProps
  | SearchBarAndroidProps
  | SearchBarIosProps;
```
This Works for me, Can you confirm @djMax

**Comment by rush86999** [comment]
i can confirm it works and platform prop becomes a required prop

**Comment by dancixx** [comment]
I have the same issue. :(

**Comment by arpitBhalla** [comment]
> I have the same issue. :(

Did you checked with `next` version?

**Comment by dancixx** [comment]
I have 3.4.2 now.

**Comment by arpitBhalla** [comment]
Will be resolved soon #3121

**Comment by dancixx** [comment]
Thank you. :)

**Comment by harshal-moneylion** [comment]
What is the release plan for the same?
been 25 days since the last activity here. And no new release has been done.

**Comment by haythamhakim** [comment]
Any updates on this?

**Comment by Estevete** [comment]
Do we have any status for this issue?

**Comment by yeocak** [comment]
Still got error with Typescript :/

**Comment by mortocks** [comment]
Same version 3.4.2 with typescript

**Comment by JimTeva** [comment]
Looks like @arpitBhalla has pushed some fixes to the branch #next but this is not yet "released".
- Remove current react-native-elements
```sh
yarn remove react-native-elements
or
npm uninstall react-native-elements
```
- add the branch #next
```sh
yarn add https://github.com/react-native-elements/react-native-elements.git#next
or
npm install --save https://github.com/react-native-elements/react-native-elements.git#next
```

**Comment by JimTeva** [comment]
@arpitBhalla provided a way to use the bleeding edge version of the library that fixes those types problem (see https://github.com/react-native-elements/react-native-elements/issues/3261)
```sh
yarn add  https://github.com/react-native-elements/react-native-elements#dist
```
Just tested, it works!

**Comment by Oize** [comment]
> ```shell
> yarn add  https://github.com/react-native-elements/react-native-elements#dist
> ```

Doesn't work for me. Yarn throws me "Invalid descriptor" error. Does anyone know a workaround?

**Comment by JimTeva** [comment]
@Oize or use patch-package library to patch yourself while the errors are getting fixed on the repo

**Comment by myxmaster** [comment]
Still not fixed in 3.4.3. I am trying to bind a method with signature ```(query: string) => void``` but get this error:

    Type '(query: string) => void' is not assignable to type '((text: string) => void) & ((text: string) => void) & (() => any) & (() => any) & (() => any)'

It only works with parameterless methods.
